### ISSUE TITLE
Global Commands

### ISSUE THREAD

**Issue by mweels** [efacilitation/eventric#46]
Does Eventric have any concept of global commands that could run on every command call?   Many times this capability would be used for logging, and/or authentication?

I reviewed the code, and didn't see anything, I wanted to check before I attempt a pull request or discuss an approach. (if your accepting of either).

Thank you in advance,
Marty

**Comment by alexlawrence** [comment]
If I understand you correctly this is about what you call a middleware mechanism in the JavaScript field, in other fields this may be called Aspects as in AOP. Am I correct? I find the wording command in "global commands" ambiguous.

**Comment by mweels** [comment]
Yes, middleware is correct.

**Comment by alexlawrence** [comment]
You are correct. There is no such feature. We considered to add it but we see no use for it in our projects using eventric. We implemented authorization in one project and decided against solving with a middleware mechanism.

You mentioned logging and authentication. I think as long as it is code you implement you could use a standalone mechanism for realising middleware. 

An example for logging:

```coffeescript
decorateCommandHandlersWithLogging = (commandHandlers) ->
  decoratedCommandHandlers = {}
  Object.keys(commandHandlers).forEach (commandName) ->
    decoratedCommandHandlers[commandName] = (params) ->
      console.log "executing command #{commandName} with #{JSON.stringify params}"
      return commandHandlers[commandName].apply @, params 
  return decoratedCommandHandlers
 ```

Something similar could be used for authentication. Maybe there are even libraries for decorating functions/objects.

What do you think? Do you see any advantage of implementing this directly in eventric?

**Comment by mweels** [comment]
Yes, I will just create the stack of commands outside of the eventric.  I guess the only positive to having it inside would be if multiple adapters were added to instantiate commands they would have the capability across clients.  Until then, I’ll keep it outside.

On Jul 9, 2016, at 1:32 PM, Alex Lawrence <notifications@github.com<mailto:notifications@github.com>> wrote:


You are correct. There is no such feature. We considered to add it but we see no use for it in our projects using eventric. We implemented authorization in one project and decided against solving with a middleware mechanism.

You mentioned logging and authentication. I think as long as it is code you implement you could use a standalone mechanism for realising middleware.

An example for logging:

decorateCommandHandlersWithLogging = (commandHandlers) ->
  decoratedCommandHandlers = {}
  Object.keys(commandHandlers).forEach (commandName) ->
    decoratedCommandHandlers[commandName] = (params) ->
      console.log "executing command #{commandName} with #{JSON.stringify params}"
      return commandHandlers[commandName].apply @, params
  return decoratedCommandHandlers

Something similar could be used for authentication. Maybe there are even libraries for decorating functions/objects.

What do you think? Do you see any advantage of implementing this directly in eventric?
